bpo-43112: detect musl as a separate SOABI

ncopa opened the pull request:
musl libc and gnu libc are not ABI compatible so we need set different
SOABI for musl and not simply assume that all linux is linux-gnu.

Replace linux-gnu with the detected os for the build from config.guess
for linux-musl*.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-43112](https://bugs.python.org/issue43112) -->
https://bugs.python.org/issue43112
<!-- /issue-number -->

Automerge-Triggered-By: GH:brettcannon

ncopa committed:
bpo-43112: detect musl as a separate SOABI

musl libc and gnu libc are not ABI compatible so we need set different
SOABI for musl and not simply assume that all linux is linux-gnu.

Replace linux-gnu with the detected os for the build from config.guess
for linux-musl*.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

tiran reviewed (dismissed):
You also have to re-generate configure script with autoreconf and commit the new configure file.

ncopa committed:
regenerate configure

ncopa committed:
Revert the runstate changes in the generated configure

ncopa committed:
Fix sysconfig test to work with -musl SOABI

commit a8e2c0c0698a (bpo-43112: detect musl as a separate SOABI)
introduced a new suffix to shared libs with musl libc. Adjust the
sysconfig test accordingly.

merwok commented on Lib/test/test_sysconfig.py:
```diff
@@ -387,10 +387,13 @@ def test_triplet_in_ext_suffix(self):
-                self.assertTrue(suffix.endswith('x86_64-linux-gnu.so'), suffix)
+                self.assertTrue(suffix.endswith('x86_64-linux-gnu.so') or
+                                suffix.endswith('x86_64-linux-musl.so'),
+                                suffix)
```
`endswith` accepts a tuple of candidates, which could be nice to use here.

I’m also wondering about the usefulness of using `suffix` as a message parameter if the test fails here; maybe it could be `f'unexpected suffix {suffix}'`?

brettcannon replied:
```suggestion
                expected_suffixes = 'x86_64-linux-gnu.so', 'x86_64-linux-musl.so'
                self.assertTrue(suffix.endswith(expected_suffixes),
                                f'unexpected suffix {suffix!r}'))
```

brettcannon reviewed (changes requested):
I'm agreeing with @merwok and their suggestion on how to touch up the tests.

brettcannon commented on Lib/test/test_sysconfig.py:
```diff
@@ -387,10 +387,13 @@ def test_triplet_in_ext_suffix(self):
-                                suffix.endswith('x86_64-linux-gnux32.so'),
+                                suffix.endswith('x86_64-linux-gnux32.so') or
+                                suffix.endswith('i386-linux-musl.so'),
                                 suffix)
```
```suggestion
                                f'unexpected suffix {suffix!r}')
```

brettcannon replied:
Due to the deleted line I couldn't make a complete suggestion on using a tuple for `str.endswitch()` like I did below.

brettcannon committed:
Update Lib/test/test_sysconfig.py

brettcannon committed:
Update Lib/test/test_sysconfig.py

brettcannon commented on Lib/test/test_sysconfig.py:
```diff
@@ -387,10 +387,13 @@ def test_triplet_in_ext_suffix(self):
         if re.match('(i[3-6]86|x86_64)$', machine):
             if ctypes.sizeof(ctypes.c_char_p()) == 4:
                 self.assertTrue(suffix.endswith('i386-linux-gnu.so') or
```
```suggestion
                expected_suffixes = ('i386-linux-gnu.so', 'x86_64-linux-gnux32.so',
                                     'i386-linux-musl.so')
```

ncopa replied:
LGTM

brettcannon commented on Lib/test/test_sysconfig.py:
```diff
@@ -387,10 +387,13 @@ def test_triplet_in_ext_suffix(self):
-                                suffix)
+                                suffix.endswith('x86_64-linux-gnux32.so') or
+                                suffix.endswith('i386-linux-musl.so'),
+                                f'unexpected suffix {suffix!r}')
```
```suggestion
```

brettcannon commented on Lib/test/test_sysconfig.py:
```diff
@@ -387,10 +387,13 @@ def test_triplet_in_ext_suffix(self):
-                self.assertTrue(suffix.endswith('x86_64-linux-gnu.so'), suffix)
+                expected_suffixes = 'x86_64-linux-gnu.so', 'x86_64-linux-musl.so'
+                self.assertTrue(suffix.endswith(expected_suffixes),
+                                f'unexpected suffix {suffix!r}'))
```
```suggestion
            self.assertTrue(suffix.endswith(expected_suffixes),
                            f'unexpected suffix {suffix!r}'))
```

uranusjr replied:
~~This seems wrong? `expected_suffxies` is only defined in this branch, moving it out of `else` would trigger `AttributeError` for the 4-byte case.~~

Oh I see, it would be read together with other suggestions.

ncopa replied:
I think there is an `SyntaxError: unmatched ')'` error.

merwok replied:
There is!  But the suggestions from Brett, taken together, amount to only defining expected_suffixes in the conditional block, and call assertTrue only once outside of the blocks.

ncopa committed:
Fix unmatched ) error in Lib/test/test_sysconfig.py

ncopa committed:
Simplify test for suffixes in Lib/test/test_sysconfig.py

merwok commented on Misc/NEWS.d/next/Build/2021-02-10-17-54-04.bpo-43112.H5Lat6.rst:
```diff
@@ -0,0 +1 @@
+Detect musl libc as a separate SOABI.
```
Maybe mention the value:

```suggestion
Detect musl libc as a separate SOABI (tagged as ``linux-musl``).
```

(adapt if it’s incorrect to call that part a tag — not sure if SOABI tag is the whole thing including arch)

ncopa committed:
Update Misc/NEWS.d/next/Build/2021-02-10-17-54-04.bpo-43112.H5Lat6.rst

Co-authored-by: Éric <merwok@netwok.org>

tianon commented on Lib/test/test_sysconfig.py:
```diff
@@ -386,11 +386,11 @@ def test_triplet_in_ext_suffix(self):
             else: # 8 byte pointer size
-                self.assertTrue(suffix.endswith('x86_64-linux-gnu.so'), suffix)
+                expected_suffixes = 'x86_64-linux-gnu.so', 'x86_64-linux-musl.so'
+        self.assertTrue(suffix.endswith(expected_suffixes),
+                        f'unexpected suffix {suffix!r}')
```
I'm totally a neophyte here, but I think the buildbot failures might be due to this?  This appears like it should've been indented to be inside the `if re.match(...` block, ala:

```suggestion
            self.assertTrue(suffix.endswith(expected_suffixes),
                            f'unexpected suffix {suffix!r}')
```

merwok replied:
ah, that would be it, only one level of dedent was expected!

autobakterie commented on configure:
```diff
@@ -5351,6 +5351,11 @@ EOF
 
 if $CPP $CPPFLAGS conftest.c >conftest.out 2>/dev/null; then
   PLATFORM_TRIPLET=`grep -v '^#' conftest.out | grep -v '^ *$' | tr -d ' 	'`
+  case "$build_os" in
```
If I'm not mistaken, `PLATFORM_TRIPLET` should refer to the target platform on which cpython will run. If that is the case, the musl libc vs. glibc decision should be based on `$host_os` rather than `$build_os` as the former is based on autoconf's `AC_CANONICAL_HOST` macro[^1] which refers to the target platform that might differ from build platform in case of cross-compilation.

For example, the decision based on `$build_os` leads to 
cpython 3.9.13 build failing[^2] when cross-compiling on AMD64 Linux with glibc for mpc8548 Linux (OpenWrt) with musl. (build runs correctly if `$host_os` is used)

[^1]: https://www.gnu.org/software/autoconf/manual/autoconf-2.68/html_node/Canonicalizing.html
[^2]: "internal configure error for the platform triplet, please file a bug report"

tiran replied:
Yes, the code is wrong. In fact the entire block should be revisited. We really should check for ``__GLIBC__`` and use a heuristic to detect musl. Could you please open a new issue for the problem?

autobakterie replied:
No problem, but there's https://github.com/python/cpython/issues/87278 already. Is yet another issue needed?

autobakterie replied:
Created https://github.com/python/cpython/issues/95855
